hack/generator/pkg/astmodel/type_name_processing_register.go

```diff
@@ -0,0 +1,100 @@
+package astmodel
+
+// TypeNameProcessingRegister is a register of TypeNames that require processing, keeping track of which names have been
+// processed and which are still outstanding. This is used, for example, for a consistency check that all TypeName
+// references can be satisfied
+type TypeNameProcessingRegister struct {
+	// known is the set of all known type names, each with reasons why it's included
+	known map[TypeName][]string
+	// completed is the set of all names that have been processed
+	completed TypeNameSet
+}
+
+// NewTypeNameProcessingRegister constructs a new empty TypeNameProcessingRegister
+func NewTypeNameProcessingRegister() *TypeNameProcessingRegister {
+	return &TypeNameProcessingRegister{
+		completed: NewTypeNameSet(),
+		known:     make(map[TypeName][]string),
+	}
+}
+
+// IsCompleted returns true if the specified TypeName is in the register and processing has been completed
+func (reg *TypeNameProcessingRegister) IsCompleted(name TypeName) bool {
+	_, found := reg.known[name]
+	return found && reg.completed.Contains(name)
+}
+
+// IsPending returns true if the specified TypeName is in the register and processing has not been completed
+func (reg *TypeNameProcessingRegister) IsPending(name TypeName) bool {
+	_, found := reg.known[name]
+	return found && !reg.completed.Contains(name)
+}
+
+// Complete marks the specified name as processing complete
+// As a side effect, the name will be added to the register if it is missing
+func (reg *TypeNameProcessingRegister) Complete(name TypeName) {
+	if _, found := reg.known[name]; !found {
+		reg.known[name] = []string{}
+	}
+
+	reg.completed.Add(name)
+}
+
+// Reset marks the specified name as unprocessed (ie not complete)
+// As a side effect, the name will be added to the checklist if it is missing
```

Comment:
minor: What is "the checklist" - possibly prefer talking about it in terms of completed/known? I think you mean added to the list of completed types.

Reply:
oh, no "the checklist" is the types we know about but haven't necessarily confirmed yet?